src/build_system.ml

```diff
@@ -704,58 +713,63 @@ and static_deps t build =
 and start_rule t _rule =
   t.hook Rule_started
 
-and setup_copy_rules t ~ctx_dir ~non_target_source_files =
-  Path.Source.Set.iter non_target_source_files ~f:(fun path ->
-    let ctx_path = Path.append_source ctx_dir path in
-    let build = Build.copy ~src:(Path.source path) ~dst:ctx_path in
-    compile_rule t (Pre_rule.make build ~context:None ~env:None
-                      ~info:Source_file_copy))
+and create_copy_rules ~ctx_dir ~non_target_source_files =
+  List.map
+    (Path.Source.Set.to_list non_target_source_files)
+    ~f:(fun path ->
+      let ctx_path = Path.append_source ctx_dir path in
+      let build = Build.copy ~src:(Path.source path) ~dst:ctx_path in
+      (Pre_rule.make build ~context:None ~env:None
+         ~info:Source_file_copy))
+
+and compile_rules ~dir t rules =
+  List.map rules ~f:(compile_rule t)
+  |> List.concat_map ~f:(fun (targets, rule) ->
+    assert (Path.(=) dir rule.Internal_rule.dir);
+    List.map (Path.Set.to_list targets) ~f:(fun target -> (target, rule)))
+  |> Path.Map.of_list_reducei ~f:rule_conflict
 
-and load_dir   t ~dir = ignore (load_dir_and_get_targets t ~dir : Path.Set.t)
 and load_dir_and_produce_its_rules t ~dir =
-  let loaded = load_dir_and_get_rules_and_targets t ~dir in
-  Rules.produce loaded.rules_produced
-
-and targets_of t ~dir =         load_dir_and_get_targets t ~dir
-
-and load_dir_and_get_targets t ~dir =
-  (load_dir_and_get_rules_and_targets t ~dir).Loaded.targets
-
-and load_dir_and_get_rules_and_targets t ~dir : Loaded.t =
-  match get_dir_status t ~dir with
-  | Some Failed_to_load -> raise Already_reported
-
-  | Some (Loaded res) -> res
-
-  | Some (Forward dir') ->
-    load_dir t ~dir:dir';
-    begin match get_dir_status t ~dir with
-    | Some (Loaded res) -> res
-    | _ -> assert false
-    end
-
-  | Some Collecting_rules ->
-    die "recursive dependency between directories:\n    %s"
-      (String.concat ~sep:"\n--> "
-         (List.map t.load_dir_stack ~f:Utils.describe_target))
-
-  | None ->
-    Path.Table.add t.dirs dir Collecting_rules;
-    t.load_dir_stack <- dir :: t.load_dir_stack;
-
-    try
+  let loaded = load_dir t ~dir in
+  match loaded with
+  | Non_build _ -> ()
+  | Build loaded ->
+    Rules.produce loaded.rules_produced
+
+and targets_of t ~dir = match load_dir t ~dir with
+  | Non_build targets -> targets
+  | Build { targets_here; _ } ->
+    Path.Set.of_list (Path.Map.keys targets_here)
+
+and load_dir_and_get_buildable_targets t ~dir =
+  let loaded = load_dir t ~dir in
+  match loaded with
+  | Non_build _ -> Path.Map.empty
+  | Build { targets_here; _ } -> targets_here
+
+and load_dir t ~dir : Loaded.t =
+  Memo.exec t.dirs dir
+and
+  load_dir_impl t ~dir : Loaded.t =
```

Comment:
typo maybe? the `and` is on the line above.

Reply:
fixed